backend/app/Http/Actions/OAuth/HandleProviderCallbackAction.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Http\Actions\OAuth;
+
+use Laravel\Socialite\Facades\Socialite;
+use Illuminate\Support\Str;
+use App\Http\Controllers\Controller;
+use App\Models\User;
+use Illuminate\Support\Facades\Auth;
+
+/**
+ * @todo これはまったく分離できていないので後で分離する
+ * @todo このクッキー方式だとhttponly属性にできないというセキュリティリスクありのため本番では変更する
+ */
+final class HandleProviderCallbackAction extends Controller
+{
+    /**
+     * ソーシャルログイン処理
+     * @return App\User
+     */
+    public function __invoke()
+    {
+        /** @todo statelessメソッドを使用すると、セッション状態の確認を無効にできます。これは、クッキーベースのセッションを利用しないステートレスAPIに、ソーシャル認証を追加する場合に有用です。 */
+        $googleUser = Socialite::driver('google')->stateless()->user();
+
+        $user  = User::where('google_id', $googleUser->id)->first();
+        $token = Str::random(80);
+
+        if ($user) {
+            $user->update([
+                'name' => $googleUser->name,
+                'email' => $googleUser->email,
+                'email_verified_at' => now(),
+                'profile_photo_path' => $googleUser->avatar,
+                'login_provider' => 1,
+                'google_token' => $googleUser->token,
+                'google_refresh_token' => $googleUser->refreshToken,
+                'api_token' => hash('sha256', $token),
+            ]);
+        } else {
+            $user = User::create([
+                'name' => $googleUser->name,
+                'email' => $googleUser->email,
+                'email_verified_at' => now(),
+                'profile_photo_path' => $googleUser->avatar,
+                'login_provider' => 1,
+                'google_id' => $googleUser->id,
+                'google_token' => $googleUser->token,
+                'google_refresh_token' => $googleUser->refreshToken,
+                'api_token' => hash('sha256', $token),
+            ]);
+        }
+
+        Auth::login($user);
```

Comment:
**[phpstan]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Parameter #1 $user of static method Illuminate\Auth\SessionGuard::login() expects Illuminate\Contracts\Auth\Authenticatable, App\Models\User given.